### ISSUE TITLE
Failure Recovery Issue with PublishAsync

### ISSUE THREAD

**Issue by cocowalla** [pardahlman/RawRabbit#143]
When using `PublishAsync` and the broker goes down, I get `System.Exception`:

```
Unable to retreive channel. All existing channels are closed and none of them are recoverable.
   at RawRabbit.Channel.ChannelFactory.EnsureRequestsAreHandled()
   at RawRabbit.Channel.ChannelFactory.GetChannelAsync()
   at RawRabbit.Operations.Publisher`1.PublishAsync[TMessage](TMessage message, Guid globalMessageId, PublishConfiguration config)
   at RawRabbit.Common.BaseBusClient`1.PublishAsync[T](T message, Guid globalMessageId, Action`1 configuration)
   at MyNamespace.MyProducer.Send(Object state)
```
This is fine, and I can handle retries myself, but it would be nice if something less generic than `System.Exception` was thrown.

**Comment by pardahlman** [comment]
@cocowalla, agreed! We'll definitely change the exception type to something more specific in an upcoming version.

However, your problem is a bit intriguing, as error lost connection is handled by default, if you have `AutomaticRecovery` set to true in your `RawRabbitConfiguration` (which it is by default).

The `ChannelFactory` [explicitly tries to recover the lost connection](https://github.com/pardahlman/RawRabbit/blob/stable/src/RawRabbit/Channel/ChannelFactory.cs#L279), with only exception if the `Application` is the initiator of the shutdown (which is the case when you try to shut down the application). Could you provide me with a step-by-step instruction on how to reproduce this?

**Comment by pardahlman** [comment]
Something that you could try, that might mitigate the problem, is to set the [`InitialChannelCount`](https://github.com/pardahlman/RawRabbit/blob/stable/src/RawRabbit/Channel/ChannelFactory.cs#L301) on the `ChannelFactoryConfiguration`. You can do this by register you own instance of it in the [IoC registration](https://github.com/pardahlman/RawRabbit#dependecy-injection).

**Comment by cocowalla** [comment]
Actually, when the connection recovers is seems to silently not work. On the next attempt to `PublishAsync` I see `[BaseBusClient`1]: Publishing message...`, but then no no logs about acks. In the RabbitMQ management portal I can also see that no messages are being received. As you suggested, I tried setting `InitialChannelCount` to `1` and `2`, but to no effect.

I double-checked, and I am using `AutomaticRecovery` and also the default implementation of `ChannelFactory`.

I'm working with CoreCLR on Windows, and I'm simply doing:
`bus.PublishAsync(message, configuration: config => config.WithExchange(e => e.WithName("my_exchange")));`

Where `bus` is an instance of `IBusClient` created by my IoC container (Structuremap). If it helps, my config looks like:

```
Username = null, // Username is CN from the SSL certificate
Password = null,
Ssl = sslOptions,
Port = 5671,
Hostnames = "127.0.0.1"),
VirtualHost = "/",
Exchange = new GeneralExchangeConfiguration { Type = ExchangeType.Topic, Durable = true, AutoDelete = false },
Queue = new GeneralQueueConfiguration { Durable = true, AutoDelete = false },
AutomaticRecovery = true,
TopologyRecovery = true,
AutoCloseConnection = true,
PersistentDeliveryMode = true,
RequestTimeout = TimeSpan.FromSeconds(10),
PublishConfirmTimeout = TimeSpan.FromSeconds(5),
RecoveryInterval = TimeSpan.FromSeconds(10)
```

I doubt it makes any difference, but I'm using a custom `IConnectionFactory` implementation that inherits from `ConnectionFactory`, because I'm using certificate authentication and need to change `ConnectionFactory.AuthMechanisms` to support RabbitMQ's 'EXTERNAL' authentication mechanism.

**Comment by pardahlman** [comment]
My guess is that when you try to publish a second message after the channel failure, the task return from `GetChannelAsync` never completes, as it is waiting for a open, available channel to complete. This is also the reason why I think a larger channel pool might help. The log message from `BaseBusClient` is [in an initial stage of the publish](https://github.com/pardahlman/RawRabbit/blob/stable/src/RawRabbit/Common/BaseBusClient.cs#L47), before channels are retrieved etc.

As for the potential fix, I re-checked the code, and you need to set [`MaxChannelCount` to 2 (or higher)](https://github.com/pardahlman/RawRabbit/blob/stable/src/RawRabbit/Channel/ChannelFactory.cs#L55) in order to create a larger channel pool.

I'm not 100% sure I follow your steps of reproductions, do you get this error each time you publish? Or just on the occasions when you lost your connection to the broker? Could I reproduce by instantiating the client, shutting down the broker and then try to publish the message?

**Comment by cocowalla** [comment]
I just tried setting `MaxChannelCount` to 10, but have the same issue.

Steps to reproduce, using code that tries to publish a message every 10 seconds.

1. Connect to broker
2. Publish a message and receive an ack
3. Kill the broker
4. Try to publish a message
5. `Get System.Exception`: "Unable to retrieve channel..."
6. Restart the broker
7. Message logged "Channel '1' is recovered..."
8. Continue trying to publish a message every 10s
9. Each time a message is logged about the message being sent, but no ack received - message is never sent

This is how I'm publishing:

```
try
{
    this.bus.PublishAsync(message, configuration: 
        config => config.WithExchange(e => e.WithName("my_exchange"))).Wait(TimeSpan.FromSeconds(1));
}
catch (Exception ex)
{
    Log.Warning("Unable to send message: {Message}", ex.Message);
}
```


Interestingly, if I kill the broker *before* the first message is published, everything seems to recover OK after restarting the broker.

**Comment by pardahlman** [comment]
Ah! I'm pretty sure that I found the problem. When [throwing the exception](https://github.com/pardahlman/RawRabbit/blob/stable/src/RawRabbit/Channel/ChannelFactory.cs#L206) `_processingRequests` is still set to `true`. This means that the next time you are asking for a channel, the worker thread believes that it does the job. I'm pretty sure that if you would change `_processingRequests` to false before throwing, the recovery should work. Care to try this and see if it works?

**Comment by cocowalla** [comment]
I'll give this a try tonight and let you know the results!

**Comment by cocowalla** [comment]
I just tried this, but get the same result, 'silent' failure - on the next attempt to `PublishAsync` I see `[BaseBusClient```1]: Publishing message...`, but then no no logs about acks :(

**Comment by pardahlman** [comment]
Ok, let me try to reproduce this locally and see if I can make sense of it.

**Comment by pardahlman** [comment]
@cocowalla - I can not reproduce this. I followed your steps and I get to 6., but after restarting the broker, messages are published (which I can confirm by looking in the mgmt plugin).

What I noticed is that I get the `BrokerUnreachableExcpetion`, rather than the other exception. How do you instantiate the bus client? What version of ERL and RabbitMq broker are you running?

Here's my stack strace
```
[11:24:43] [$11] [INFO] [ChannelFactory]: Channel '2' is closed. Removing it from pool.
[11:24:49] [$10] [ERROR] [ChannelFactory]: Unable to connect to broker, Exception:
 RabbitMQ.Client.Exceptions.BrokerUnreachableException: None of the specified endpoints were reachable ---> RabbitMQ.Client.Exceptions.ConnectFailureException: Connection failed ---> System.Net.Sockets.SocketException: No connection could be made because the target machine actively refused it 127.0.0.1:5672
   at System.Net.Sockets.Socket.DoConnect(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Connect(EndPoint remoteEP)
   at System.Net.Sockets.Socket.Connect(IPAddress address, Int32 port)
   at RabbitMQ.Client.TcpClientAdapter.<ConnectAsync>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at RabbitMQ.Client.Impl.TaskExtensions.<TimeoutAfter>d__0.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at RabbitMQ.Client.Impl.SocketFrameHandler.Connect(ITcpClient socket, AmqpTcpEndpoint endpoint, Int32 timeout)
   --- End of inner exception stack trace ---
   at RabbitMQ.Client.EndpointResolverExtensions.SelectOne[T](IEndpointResolver resolver, Func`2 selector)
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.Init(IEndpointResolver endpoints)
   at RabbitMQ.Client.ConnectionFactory.CreateConnection(IEndpointResolver endpointResolver, String clientProvidedName)
   --- End of inner exception stack trace ---
   at RabbitMQ.Client.ConnectionFactory.CreateConnection(IEndpointResolver endpointResolver, String clientProvidedName)
   at RabbitMQ.Client.ConnectionFactory.CreateConnection(IList`1 hostnames, String clientProvidedName)
   at RabbitMQ.Client.ConnectionFactory.CreateConnection(IList`1 hostnames)
   at RawRabbit.Channel.ChannelFactory..ctor(IConnectionFactory connectionFactory, RawRabbitConfiguration config, ChannelFactoryConfiguration channelConfig) in ChannelFactory.cs:line 34
   [11:25:27] [$10] [DEBUG] [ChannelFactory]: Initiating 0 channels.
[11:25:27] [$10] [INFO] [ChannelFactory]: Channel scaling is disabled.
[11:25:27] [$10] [INFO] [BaseBusClient`1]: BusClient initialized.
Publishing Message.
[11:25:27] [$10] [DEBUG] [BaseBusClient`1]: Publishing message 'ConnectionMessage' on exchange 'connectiontest' with routing key connectionmessage.
[11:25:27] [$10] [DEBUG] [TopologyProvider]: Start processing topology work.
[11:25:27] [$10] [INFO] [TopologyProvider]: Declaring exchange 'connectiontest'.
[11:25:27] [$10] [DEBUG] [ChannelFactory]: Existing connection is open and will be used.
[11:25:27] [$10] [DEBUG] [TopologyProvider]: Done processing topology work.
[11:25:27] [$10] [DEBUG] [ChannelFactory]: Begining to process 'GetChannel' requests.
[11:25:27] [$10] [DEBUG] [ChannelFactory]: Existing connection is open and will be used.
[11:25:27] [$11] [INFO] [ChannelFactory]: Channel '2' has been created.
[11:25:27] [$10] [DEBUG] [ChannelFactory]: 'GetChannel' has been processed.
[11:25:27] [$16] [INFO] [PublishAcknowledger]: Setting 'Publish Acknowledge' for channel '2'
[11:25:27] [$21] [INFO] [PublishAcknowledger]: Recieved ack for 1/2 with multiple set to 'False'
[11:25:27] [$21] [DEBUG] [PublishAcknowledger]: Disposed ack timer for 1/2
Sleeping...
[11:25:27] [$21] [DEBUG] [PublishAcknowledger]: Successfully confirmed publish 1/2
[11:25:29] [$11] [INFO] [TopologyProvider]: Disposing topology channel (if exists).
```

**Comment by cocowalla** [comment]
I get the same as you *if* no messages have been published with the IBusClient before the broker dies - are you sure something was published beforehand in your test, and you were using the same IBusClient?

I'm using the latest version of RabbitMQ with Erlang 18.3 (there seems to be SSL/TLS issues with 19.x).

I shouldn't think it matters, but we're you testing over an SSL connection?

**Comment by pardahlman** [comment]
Yes, I published 2-3 messages before killing the broker and yes I published with the same client. I used the default settings (by creating the client from `BusClientFactory.CreateDefault()`, so no TLS.

Come to think about it I did run my tests in a "classic" console app (not .NET Core). What target framework are you using? `netstandard1.5`? There could perhaps be some compiler directives in `RabbitMQ.Client` that is the cause of the issue.

**Comment by cocowalla** [comment]
I'm targeting. NET Standard 1.6. I've only been testing with Windows 7 so far, but will be supporting Windows 7+ and various flavours of Linux.

I'll give it a try later on without TLS, using `BusClientFactory.CreateDefault` and let you know how I get on.

**Comment by pardahlman** [comment]
Looking forward to it 👍 . I've got my mind set for 2.0, but I'm more than happy to accept a PR if you find the cause of the problem.

**Comment by cocowalla** [comment]
I tried it without SSL, and using the same config as `BusClientFactory.CreateDefault`, but I get the same result :(

Any idea when 2.0 is going to be feature-complete?

**Comment by pardahlman** [comment]
Roger that. It would be interesting to see how your setup would perform in a "classic" console app under 4.5.1.

As for 2.0... A lot of the pieces are coming together pretty fast, but there are still a few things left to do where Channel Management #119 is a priority. I want to overlook the resilience of Consumer as well, so that it will re-establish the consumers nicely in case of longer network downtime. #146, or a first stab if it, needs to be in place to. After a period of bench marking, I'll probably release a pre-release of it, but I think that will probably be mid-January.

**Comment by cocowalla** [comment]
I just did a test with a console app under 4.5.1 - same problem, exactly as with .NET Core:  [RawRabbitTest.zip](https://github.com/pardahlman/RawRabbit/files/589111/RawRabbitTest.zip)


Looking forward to 2.0 :)

**Comment by pardahlman** [comment]
Ok @cocowalla - I found the problem. The issue was that `GetChannelAsync` did not verify that the connection itself was open. Once I realized that, the fix was pretty easy. In the case of a lost connection, the channel factory returns a `Task<IModel>` that will be completed after connection recovery (if recovery is not disabled from config). That means that publishes initiated under the downtime will be published to the broker once the connection is reestablished. I also took the liberty to update the logging from `BaseBusClient` to indicate that a publish is _initialized_.

This is the output when running your sample code. For each published message when the broker is running you'll see this
```
[DEBUG] [BaseBusClient`1]: Initiating publish for message 'HeartbeatMessage' on exchange 'heartbeat' with routing key heartbeatmessage.
[DEBUG] [ChannelFactory]: Begining to process 'GetChannel' requests.
[DEBUG] [ChannelFactory]: 'GetChannel' has been processed.
[INFO] [PublishAcknowledger]: Recieved ack for 5/2 with multiple set to 'False'
[DEBUG] [PublishAcknowledger]: Disposed ack timer for 5/2
[DEBUG] [PublishAcknowledger]: Successfully confirmed publish 5/2
```
When the broker is shut down, you only see a bunch of these
```
[DEBUG] [BaseBusClient`1]: Initiating publish for message 'HeartbeatMessage' on exchange 'heartbeat' with routing key heartbeatmessage.
[DEBUG] [BaseBusClient`1]: Initiating publish for message 'HeartbeatMessage' on exchange 'heartbeat' with routing key heartbeatmessage.
[DEBUG] [BaseBusClient`1]: Initiating publish for message 'HeartbeatMessage' on exchange 'heartbeat' with routing key heartbeatmessage.
[DEBUG] [BaseBusClient`1]: Initiating publish for message 'HeartbeatMessage' on exchange 'heartbeat' with routing key heartbeatmessage.
```
And when connection is recovered you see this
```
Topology recovery exception: RabbitMQ.Client.TopologyRecoveryException: Caught an exception while recovering exchange heartbeat: Already closed: The AMQP operation was interrupted: AMQP close-reason, initiated by Application, code=200, text="Goodbye", classId=0, methodId=0, cause= ---> RabbitMQ.Client.Exceptions.AlreadyClosedException: Already closed: The AMQP operation was interrupted: AMQP close-reason, initiated by Application, code=200, text="Goodbye", classId=0, methodId=0, cause=
   at RabbitMQ.Client.Impl.SessionBase.Transmit(Command cmd)
   at RabbitMQ.Client.Impl.ModelBase.ModelRpc(MethodBase method, ContentHeaderBase header, Byte[] body)
   at RabbitMQ.Client.Framing.Impl.Model._Private_ExchangeDeclare(String exchange, String type, Boolean passive, Boolean durable, Boolean autoDelete, Boolean internal, Boolean nowait, IDictionary`2 arguments)
   at RabbitMQ.Client.Impl.ModelBase.ExchangeDeclare(String exchange, String type, Boolean durable, Boolean autoDelete, IDictionary`2 arguments)
   at RabbitMQ.Client.Impl.RecordedExchange.Recover()
   at RabbitMQ.Client.Framing.Impl.AutorecoveringConnection.RecoverExchanges()
   --- End of inner exception stack trace ---
[INFO] [ChannelFactory]: Connection has been recovered. Starting channel processing.
```
Where the first part is from the underlying library, see https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/254. If you create a queue and bind it to the exchange correctly, you can verify that the number of messages in that queue is the same as number of calls to `PublishAsync`.

**Comment by cocowalla** [comment]
Superb, I should be able to test this out tonight or tomorrow!

**Comment by cocowalla** [comment]
Ah, I'm really sorry, I totally forgot about this until I saw the notification about it being closed. I've tried this out, and it does indeed work well!

Something of note is that messages 'published' while the broker is stopped seem to be buffered and sent when it is back up again. I can see how this is desirable in many cases, but not all. Is there some way to control this behaviour (either to limit the buffer size or to discard messages that couldn't be sent)?
